core/modules/publisher-handler.js

```diff
@@ -0,0 +1,181 @@
+/*\
+title: $:/core/modules/publisherhandler.js
+type: application/javascript
+module-type: global
+
+The publisher manages publishing extracts of wikis as external files
+
+\*/
+(function(){
+
+/*jslint node: true, browser: true */
+/*global $tw: false */
+"use strict";
+
+var PUBLISHING_MODAL_TITLE = "$:/language/Publishing/Modal";
+
+/*
+Instantiate the publisher manager with the following options
+wiki: wiki object to be used
+commander: commander object to be used for output
+*/
+function PublisherHandler(options) {
+	this.wiki = options.wiki;
+	this.commander = options.commander;
+}
+
+/*
+Publish a job
+
+jobTitle: title of tiddler containing details of the job
+callback: completion callback invoked callback(err)
+options: Include:
+
+commander: commander object associated with publishing under Node.js
+variables: hashmap of variables to be passed to renderings
+*/
+PublisherHandler.prototype.publish = function(jobTitle,callback,options) {
+	if(jobTitle) {
+		var job = new PublishingJob(jobTitle,this,options);
+		job.publish(callback);
+	}
+};
+
+function PublishingJob(jobTitle,publisherHandler,options) {
+	options = options || {};
+	// Save params
+	this.jobTitle = jobTitle;
+	this.publisherHandler = publisherHandler;
+	this.commander = options.commander;
+	this.publishVariables = options.variables || Object.create(null);
+}
+
+/*
+Start publishing
+*/
+PublishingJob.prototype.publish = function(callback) {
+	var self = this;
+	// Get the job tiddler and check it is enabled
+	this.jobTiddler = this.publisherHandler.wiki.getTiddler(this.jobTitle);
+	if(this.jobTiddler && this.jobTiddler.fields.enabled === "yes") {
+		// Get the list of tiddlers to be exported, defaulting to all non-system tiddlers
+		this.exportList = this.publisherHandler.wiki.filterTiddlers(this.jobTiddler.fields["export-filter"] || "[!is[system]]");
+		// Get publisher
+		this.publisher = this.getPublisher(this.jobTiddler.fields.publisher);
+		if(this.publisher) {
+			// Get the sitemap
+			this.sitemap = new $tw.Sitemap(this.jobTiddler.fields.sitemap,{
+				wiki: this.publisherHandler.wiki,
+				variables: this.publishVariables
+			});
+			this.sitemap.load();
+			// Get the output operations
+			this.operations = this.sitemap.getAllFileDetails(this.exportList);
+			// Display the progress modal
+			if($tw.modal) {
+				this.progressModal = $tw.modal.display(PUBLISHING_MODAL_TITLE,{
```

Comment:
We may need a means for multiple jobs to use the same modal, if one already exists. Currently with multiple enabled jobs we get one modal over another.